File: project/jimmer-sql/src/test/java/org/babyfish/jimmer/sql/model/filter/File.java
package org.babyfish.jimmer.sql.model.filter;

import org.babyfish.jimmer.sql.*;
import org.jetbrains.annotations.Nullable;

import java.util.List;

@Entity
@KeyUniqueConstraint(isNullNotDistinct = true)
public interface File {

    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    long id();

    @Key
    String name();

    @ManyToOne
    @Key
    @Nullable
    @OnDissociate(DissociateAction.DELETE)
    File parent();

    @OneToMany(mappedBy = "parent", orderedProps = @OrderedProp("id"))
    List<File> childFiles();

    @ManyToMany
    @JoinTable(deletedWhenEndpointIsLogicallyDeleted = true)
    List<User> users();
}

File: project/jimmer-sql-kotlin/src/test/kotlin/org/babyfish/jimmer/sql/kt/model/filter/File.kt
package org.babyfish.jimmer.sql.kt.model.filter

import org.babyfish.jimmer.sql.*

@Entity
interface File {

    @Id
    val id: Long

    @Key
    val name: String

    @ManyToOne
    @Key
    @OnDissociate(DissociateAction.DELETE)
    val parent: File?

    @OneToMany(mappedBy = "parent", orderedProps = [OrderedProp("id")])
    val childFiles: List<File>

    @ManyToMany
    val users: List<User>
}